queue/driver_redis.go

```diff
@@ -0,0 +1,189 @@
+package queue
+
+import (
+	"context"
+	"strconv"
+	"time"
+
+	"github.com/redis/go-redis/v9"
+
+	contractsqueue "github.com/goravel/framework/contracts/queue"
+	"github.com/goravel/framework/support/json"
+)
+
+type redisData struct {
+	Signature string               `json:"signature"`
+	Args      []contractsqueue.Arg `json:"args"`
+	Attempts  uint                 `json:"attempts"`
+}
+
+type Redis struct {
+	connection string
+	retryAfter uint
+	client     *redis.Client
+	ctx        context.Context
+}
+
+func NewRedis(connection string, client *redis.Client) *Redis {
+	return &Redis{
+		connection: connection,
+		retryAfter: 60,
+		client:     client,
+		ctx:        context.Background(),
+	}
+}
+
+func (r *Redis) ConnectionName() string {
+	return r.connection
+}
+
+func (r *Redis) DriverName() string {
+	return DriverRedis
+}
+
+func (r *Redis) Push(job contractsqueue.Job, args []contractsqueue.Arg, queue string) error {
+	payload, err := r.jobToJSON(job.Signature(), args)
+	if err != nil {
+		return err
+	}
+
+	return r.client.RPush(r.ctx, queue, payload).Err()
+}
+
+func (r *Redis) Bulk(jobs []contractsqueue.Jobs, queue string) error {
+	pipe := r.client.Pipeline()
+
+	for _, job := range jobs {
+		payload, err := r.jobToJSON(job.Job.Signature(), job.Args)
+		if err != nil {
+			return err
+		}
+
+		if job.Delay > 0 {
+			delayDuration := time.Duration(job.Delay) * time.Second
+			pipe.ZAdd(r.ctx, queue+":delayed", redis.Z{
+				Score:  float64(time.Now().Add(delayDuration).Unix()),
+				Member: payload,
+			})
+		} else {
+			pipe.RPush(r.ctx, queue, payload)
+		}
+	}
+
+	_, err := pipe.Exec(r.ctx)
+	return err
+}
+
+func (r *Redis) Later(delay uint, job contractsqueue.Job, args []contractsqueue.Arg, queue string) error {
+	payload, err := r.jobToJSON(job.Signature(), args)
+	if err != nil {
+		return err
+	}
+
+	delayDuration := time.Duration(delay) * time.Second
+	return r.client.ZAdd(r.ctx, queue+":delayed", redis.Z{
+		Score:  float64(time.Now().Add(delayDuration).Unix()),
+		Member: payload,
+	}).Err()
+}
+
+func (r *Redis) Pop(queue string) (contractsqueue.Job, []contractsqueue.Arg, error) {
+	if err := r.migrateDelayedJobs(queue); err != nil {
+		return nil, nil, err
+	}
+
+	result, err := r.client.BLPop(r.ctx, 1*time.Second, queue).Result()
```

Comment:
This timeout Laravel is setting to 60 seconds. But I think 60 it too long, So I set it to 1 second.

Reply:
1 second is too short when using a remote Redis server, It's read timeout second, so I think 60 seconds is ok.